File: interfaces/python/tests/unittests/infeasibility_tests.py
# Test osqp python module
import osqp
import scipy.sparse as spspa
import scipy as sp
import numpy as np

# Unit Test
import unittest


class infeeasibility_tests(unittest.TestCase):

    def setUp(self):
        sp.random.seed(6)
        """
        Setup infeasible problem
        """

        self.opts = {'verbose': False,
                     'eps_abs': 1e-05,
                     'eps_rel': 1e-05,
                     'scaling': True,
                     'scaling_norm': 2,
                     'scaling_iter': 3,
                     'rho': 1.6,
                     'alpha': 1.6,
                     'max_iter': 2500,
                     'polishing': False}

    def test_infeasibility_problem(self):

        # Simple QP problem
        sp.random.seed(4)

        self.n = 50
        self.m = 500
        # Generate random Matrices
        Pt = spspa.random(self.n, self.n)
        self.P = Pt.T.dot(Pt).tocsc()
        self.q = sp.randn(self.n)
        self.A = spspa.random(self.m, self.n).tolil()  # Lil for efficiency
        self.u = 3 + sp.randn(self.m)
        self.l = -3 + sp.randn(self.m)

        # Make random problem infeasible
        self.A[int(self.n/2), :] = self.A[int(self.n/2)+1, :]
        self.l[int(self.n/2)] = self.u[int(self.n/2)+1] + 10 * sp.rand()
        self.u[int(self.n/2)] = self.l[int(self.n/2)] + 0.5

        # Convert A to csc
        self.A = self.A.tocsc()

        self.model = osqp.OSQP()
        self.model.setup(P=self.P, q=self.q, A=self.A, l=self.l, u=self.u,
                         **self.opts)

        # Solve problem with OSQP
        res = self.model.solve()

        # Assert close
        self.assertEqual(res.info.status_val,
                         self.model.constant('OSQP_INFEASIBLE'))

    def test_infeasible_and_unbounded_problem(self):

        self.n = 2
        self.m = 4
        self.P = spspa.csc_matrix((2, 2))
        self.q = np.array([-1., -1.])
        self.A = spspa.csc_matrix([[1., -1.], [-1., 1.], [1., 0.], [0., 1.]])
        self.l = np.array([1., 1., 0., 0.])
        self.u = np.inf * np.ones(self.m)

        self.model = osqp.OSQP()
        self.model.setup(P=self.P, q=self.q, A=self.A, l=self.l, u=self.u,
                         **self.opts)

        res = self.model.solve()

        # Assert close
        self.assertEqual(res.info.status_val,
                         self.model.constant('OSQP_INFEASIBLE'))

File: interfaces/matlab/unittests/infeasibility_tests.m
classdef infeasibility_tests < matlab.unittest.TestCase
    %FEASIBILITY_TESTS Solve equality constrained feasibility problem

    properties
        P
        q
        A
        u
        l
        m,
        n
        solver
        options
        quadprog_opts
        tol
    end

    methods(TestMethodSetup)
        function setup_problem(testCase)
  
            % Setup solver options
            testCase.options = struct;
            testCase.options.verbose = 0;
            testCase.options.rho = 0.01;
            testCase.options.eps_inf = 1e-05;
            testCase.options.max_iter = 2500;



            % Setup quadprog options
            testCase.quadprog_opts.Display = 'off';
            % testCase.quadprog_opts.Algorithm = 'dual-simplex';

            % Setup tolerance
            testCase.tol = 1e-04;

        end
    end

    methods (Test)
        function test_infeasibility_problem(testCase)
            
            % Create Problem
            rng(4)
            testCase.n = 50;
            testCase.m = 500;
            Pt = sprandn(testCase.n, testCase.n, 0.6);
            testCase.P = Pt' * Pt;
            testCase.q = randn(testCase.n, 1);
            testCase.A = sprandn(testCase.m, testCase.n, 0.8);
            testCase.u = 3 + randn(testCase.m, 1);
            testCase.l = -3 + randn(testCase.m, 1);

            % Make random problem infeasible
            nhalf = floor(testCase.n/2);
            testCase.A(nhalf, :) = ...
                testCase.A(nhalf + 1, :);
            testCase.l(nhalf) = testCase.u(nhalf + 1) + 10 * rand();
            testCase.u(nhalf) = testCase.l(nhalf) + 0.5;

            
            % Setup solver
            testCase.solver = osqp;
            testCase.solver.setup(testCase.P, testCase.q, ...
                testCase.A, testCase.l, testCase.u, testCase.options);
            
            % Solve with OSQP
            results = testCase.solver.solve();

            % Solve with quadprog to double-check that it is infeasible
            % [~, ~, ~] = quadprog(testCase.P, testCase.q, ...
            %               [testCase.A; -testCase.A], ...
            %               [testCase.u; -testCase.l], ...
            %               [],[],[],[],[], testCase.quadprog_opts);

            % Check if they are close
            testCase.verifyEqual(results.info.status_val, ...
                                 testCase.solver.constant('OSQP_INFEASIBLE'), ...
                                 'AbsTol', testCase.tol)

        end
        
        
        function test_unbounded_and_infeasible(testCase)
            testCase.n = 2;
            testCase.m = 4;
            testCase.P = sparse(2,2);
            testCase.q = [-1;-1];
            testCase.A = sparse([1 -1; -1 1;1 0;0 1]);
            testCase.l = [1; 1; 0; 0];
            testCase.u = Inf(testCase.m,1);

            % Setup solver
            testCase.solver = osqp;
            testCase.solver.setup(testCase.P, testCase.q, ...
                testCase.A, testCase.l, testCase.u, testCase.options);
            
            % Solve with OSQP
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.info.status_val, ...
                                 testCase.solver.constant('OSQP_INFEASIBLE'), ...
                                 'AbsTol', testCase.tol)

        end

    end

end
